TESTS/netsocket/tls/tlssocket_echotest.cpp

```diff
@@ -0,0 +1,195 @@
+        fill_tx_buffer_ascii(tls_global::tx_buffer, BUFF_SIZE);
+
+        sent = sock->send(tls_global::tx_buffer, pkt_s);
+        if (sent < 0) {
+            printf("[Round#%02d] network error %d\n", x, sent);
+            TEST_FAIL();
+            TEST_ASSERT_EQUAL(NSAPI_ERROR_OK, sock->close());
+            delete sock;
+            return;
+        }
+
+        int bytes2recv = sent;
+        while (bytes2recv) {
+            recvd = sock->recv(&(tls_global::rx_buffer[sent - bytes2recv]), bytes2recv);
+            if (recvd < 0) {
+                printf("[Round#%02d] network error %d\n", x, recvd);
+                TEST_FAIL();
+                TEST_ASSERT_EQUAL(NSAPI_ERROR_OK, sock->close());
+                delete sock;
+                return;
+            }
+            bytes2recv -= recvd;
+        }
+        TEST_ASSERT_EQUAL(0, memcmp(tls_global::tx_buffer, tls_global::rx_buffer, sent));
+    }
+    TEST_ASSERT_EQUAL(NSAPI_ERROR_OK, sock->close());
+    delete sock;
+}
+
+void tlssocket_echotest_nonblock_receiver(void *receive_bytes)
+{
+    int bytes2recv = *(int *)receive_bytes;
+    int recvd;
+    while (bytes2recv) {
+        recvd = sock->recv(&(tls_global::rx_buffer[*(int *)receive_bytes - bytes2recv]), bytes2recv);
+        if (recvd == NSAPI_ERROR_WOULD_BLOCK) {
+            if (tc_exec_time.read() >= time_allotted) {
+                TEST_FAIL();
+                break;
+            }
+            wait(1);
+            continue;
+        } else if (recvd < 0) {
+            TEST_FAIL();
+            break;
+        }
+        bytes2recv -= recvd;
+    }
+
+    TEST_ASSERT_EQUAL(0, memcmp(tls_global::tx_buffer, tls_global::rx_buffer, *(int *)receive_bytes));
+
+    static int round = 0;
+    printf("[Recevr#%02d] bytes received: %d\n", round++, *(int *)receive_bytes);
+
+    tx_sem.release();
+
+}
+
+void TLSSOCKET_ECHOTEST_NONBLOCK()
+{
+#if MBED_CONF_NSAPI_SOCKET_STATS_ENABLE
+    int j = 0;
+    int count = fetch_stats();
+    for (; j < count; j++) {
+        TEST_ASSERT_EQUAL(SOCK_CLOSED, tls_stats[j].state);
+    }
+#endif
+    sock = new TLSSocket;
+    tc_exec_time.start();
+    time_allotted = split2half_rmng_tls_test_time(); // [s]
+
+    tlssocket_connect_to_echo_srv(*sock);
+    sock->set_blocking(false);
+    sock->sigio(callback(_sigio_handler, ThisThread::get_id()));
+
+    int bytes2send;
+    int sent;
+    int s_idx = 0;
+    Thread *thread;
+    unsigned char *stack_mem = (unsigned char *)malloc(tls_global::TLS_OS_STACK_SIZE);
+    TEST_ASSERT_NOT_NULL(stack_mem);
+
+    for (int pkt_s = pkt_sizes[s_idx]; s_idx < PKTS; ++s_idx) {
+        pkt_s = pkt_sizes[s_idx];
+        thread = new Thread(osPriorityNormal,
+                            tls_global::TLS_OS_STACK_SIZE,
+                            stack_mem,
+                            "receiver");
+        TEST_ASSERT_EQUAL(osOK, thread->start(callback(tlssocket_echotest_nonblock_receiver, &pkt_s)));
```

Comment:
New thread per packet?

Sounds a bit overkill...


We could push event to global event loop every time we receive sigio(), and that event could the read until WOULD_BLOCK is received.. but if all data is received, it could set some global event flag to indicate that all data is read.